Add k8s-specific fields to session.start events

awly opened the pull request:
Added fields:
- `k8s_container_name`
- `k8s_pod_name`
- `k8s_pod_namespace`
- `k8s_auth_users`
- `k8s_auth_groups`

Also add tx/rx fields to session.end.

Updates #3528

Commit:
Add k8s-specific fields to session.start events

Added fields:
- `k8s_container_name`
- `k8s_pod_name`
- `k8s_pod_namespace`
- `k8s_auth_users`
- `k8s_auth_groups`

Also add tx/rx fields to session.end.

awly commented:
Example `session.start`:
```
{
  "addr.local": "172.17.0.2:8443",
  "addr.remote": "[::1]:60948",
  "code": "T2000I",
  "ei": 0,
  "event": "session.start",
  "k8s_auth_groups": [
    "system:authenticated",
    "system:masters"
  ],
  "k8s_auth_users": [
    "awly"
  ],
  "k8s_container_name": "busybox",
  "k8s_pod_name": "busybox",
  "k8s_pod_namespace": "default",
  "login": "awly",
  "namespace": "default",
  "proto": "kube",
  "server_id": "7b515a99-8f38-46c2-a9a5-65012f085d12",
  "sid": "b705d5fa-e47a-4bd1-8622-358a42c0f880",
  "size": "100:100",
  "time": "2020-04-29T00:24:33.301Z",
  "uid": "bb7e15e5-e869-473a-a838-1af233468b72",
  "user": "awly"
}
```

Example `session.end`:
```
{
  "code": "T2004I",
  "ei": 5,
  "event": "session.end",
  "namespace": "default",
  "proto": "kube",
  "rx": 16,
  "sid": "b705d5fa-e47a-4bd1-8622-358a42c0f880",
  "time": "2020-04-29T00:25:00.186Z",
  "tx": 1,
  "uid": "58b2dcd1-2ee2-4662-9810-6efd66b19fb3",
  "user": "awly"
}
```

webvictim commented on lib/kube/proxy/forwarder.go:
```diff
@@ -562,18 +575,33 @@ func (f *Forwarder) exec(ctx *authContext, w http.ResponseWriter, req *http.Requ
+			// Note:
+			// - "transmitted" here means from user to pod.
+			// - "received" here means from pod to user.
+			events.DataTransmitted: trackIn.Count(),
+			events.DataReceived:    trackOut.Count() + trackErr.Count(),
```
Do you have an example of how these appear in the log output too?

awly replied:
This is how these show up:
```json
{
  "code": "T2004I",
  "ei": 5,
  "event": "session.end",
  "namespace": "default",
  "proto": "kube",
  ---> "rx": 16,
  "sid": "b705d5fa-e47a-4bd1-8622-358a42c0f880",
  "time": "2020-04-29T00:25:00.186Z",
  ---> "tx": 1,
  "uid": "58b2dcd1-2ee2-4662-9810-6efd66b19fb3",
  "user": "awly"
}
```

webvictim replied:
Thanks!

Commit:
Add audit event labels from k8s pod object on session.start

Fetch the pod object from k8s API and (if succeeded), add extra labels
to the audit event:
- pod labels
- k8s apiserver address
- k8s node the pod is running on
- k8s service account the pod is running as
- image of the container being execed into

awly commented:
PTAL, added labels from the k8s pod object:
- pod labels
- k8s cluster name
- k8s node the pod is running on
- k8s service account the pod is running as
- image of the container being execed into
```
{
  "addr.local": "172.17.0.2:8443",
  "addr.remote": "[::1]:40634",
  "code": "T2000I",
  "ei": 0,
  "event": "session.start",
  "k8s_auth_groups": [
    "system:authenticated",
    "system:masters"
  ],
  "k8s_auth_users": [
    "awly"
  ],
  "k8s_cluster": "",
  "k8s_container_image": "busybox",
  "k8s_container_name": "busybox",
  "k8s_labels": {
    "run": "busybox"
  },
  "k8s_node": "minikube",
  "k8s_pod_name": "busybox",
  "k8s_pod_namespace": "default",
  "k8s_service_account": "default",
  "login": "awly",
  "namespace": "default",
  "proto": "kube",
  "server_id": "7b515a99-8f38-46c2-a9a5-65012f085d12",
  "sid": "bb7eeb44-ba5c-461f-bf21-12a0dda770d8",
  "size": "100:100",
  "time": "2020-04-29T17:58:34.308Z",
  "uid": "35db95ca-de7b-4658-979f-fc66207e2563",
  "user": "awly"
}
```

klizhentas commented:
hey folks, can we hold back on new/modified events in master? I have migrated all events to protobuf and these changes to events that are now structured will make it harder for me to merge - so if this one can wait a bit, I'd appreciate it @russjones @awly 

You can see new events here:

https://github.com/gravitational/teleport/blob/sasha/stream/lib/events/events.proto

klizhentas commented:
@awly to make it more convenient for you, you can assign this PR to me, I will make sure to merge with stream branch after it's reviewed.

awly commented:
@klizhentas sure, as long as it lands in 4.3 (https://github.com/gravitational/teleport/issues/3528)

This is a trivial change, shouldn't be hard to merge.
I would prefer to do the merging, to dogfood the new events API and give feedback.

klizhentas commented:
Fair, one thing I spotted is the `k8s_` prefix in the events, would it be more convenient to have `K8s` as a separate protobuf structure for the future merge?

awly commented:
Yeah, separate proto is a good idea.
Except maybe `k8s_users`/`k8s_groups` could go into `UserMetadata`
